### ISSUE TITLE
Unable to add try_files to nginx config 

### ISSUE THREAD

**Issue by mfish0005** [dokku/dokku#3608]
I highly doubt this is a bug but I've been stuck on this all day.  I've followed the docs and Googled everything I can think of with no luck.

**The Problem**
I have a DO droplet running the latest version of Dokku.  One of the apps I'm running is an Angular 7 app that does *not* use hash router.  The app works fine when I navigate to `example.com`.  The problem is whenever you enter a url manually(E.G. `example.com/contact`) you are given the nginx 404 page.

Fixing this with nginx usually means adding something like `try_files $uri $uri/ /index.html =404;` to an nginx conf file.  

**What I've Tried**
1.  Adding `try_files` directly to /home/dokku/example.com/nginx.conf.  As soon as I redeploy my app the changes to `nginx.conf` are lost so this doesn't work.

2. Adding an `nginx.conf.sigil` file to the root of the project and adding `WORKDIR /app COPY nginx.conf.sigil .` to my Dockerfile.  As far as I can tell the sigil file is never used upon deployment.

3. Per the docs I tried something like this:

```
mkdir /home/dokku/example.com/nginx.conf.d/
echo 'try_files $uri $uri/ /index.html =404;' > /home/dokku/node-js-app/nginx.conf.d/redirects.conf
chown dokku:dokku /home/dokku/node-js-app/nginx.conf.d/upload.conf
service nginx reload
```

If  `/home/dokku/example.com/redirects.conf` contains `location / {try_files $uri $uri/ /index.html =404}` I of course get an error like `duplicate location / in /home/dokku/example.com/redirects.conf`

 If `/home/dokku/example.com/redirects.conf` contains simply `try_files $uri $uri/ /index.html =404` the app deploys but the problem still exists

**Conclusion**
On a regular nginx server `try_files` would fix the problem but with Dokku it doesn't.  Should these methods work?  Am I missing something here?   

Thanks in advance and amazing work btw.  Dokku works great for my API layer I was just hoping to use it to serve my client app also.

**Comment by josegonzalez** [comment]
Do you have an example app exhibiting this issue? Also, can you show the output of `dokku report`? Thanks.

**Comment by mfish0005** [comment]
@josegonzalez Thanks for the quick reply!  I can put together an example app and post it tomorrow.  
Here is my `dokku report` 
```-----> uname: Linux grayfish.io 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
-----> memory:
                     total        used        free      shared  buff/cache   available
       Mem:           1993         261         425          13        1306        1534
       Swap:             0           0           0
-----> docker version:
       Client:
        Version:           18.09.6
        API version:       1.39
        Go version:        go1.10.8
        Git commit:        481bc77
        Built:             Sat May  4 02:35:57 2019
        OS/Arch:           linux/amd64
        Experimental:      false

       Server: Docker Engine - Community
        Engine:
         Version:          18.09.6
         API version:      1.39 (minimum version 1.12)
         Go version:       go1.10.8
         Git commit:       481bc77
         Built:            Sat May  4 01:59:36 2019
         OS/Arch:          linux/amd64
         Experimental:     false
-----> docker daemon info:
       Containers: 12
        Running: 4
        Paused: 0
        Stopped: 8
       Images: 40
       Server Version: 18.09.6
       Storage Driver: overlay2
        Backing Filesystem: extfs
        Supports d_type: true
        Native Overlay Diff: true
       Logging Driver: json-file
       Cgroup Driver: cgroupfs
       Plugins:
        Volume: local
        Network: bridge host macvlan null overlay
        Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
       Swarm: inactive
       Runtimes: runc
       Default Runtime: runc
       Init Binary: docker-init
       containerd version: bb71b10fd8f58240ca47fbb579b9d1028eea7c84
       runc version: 2b18fe1d885ee5083ef9f0838fee39b62d653e30
       init version: fec3683
       Security Options:
        apparmor
        seccomp
         Profile: default
       Kernel Version: 4.15.0-52-generic
       Operating System: Ubuntu 18.04.2 LTS
       OSType: linux
       Architecture: x86_64
       CPUs: 1
       Total Memory: 1.947GiB
       Name: grayfish.io
       ID: F6S2:HC2I:QQCN:IMGS:6UZN:ZXZ6:ECJF:RYKR:RBWR:6ETQ:QJXR:7DB3
       Docker Root Dir: /var/lib/docker
       Debug Mode (client): true
       Debug Mode (server): false
       Registry: https://index.docker.io/v1/
       Labels:
       Experimental: false
       Insecure Registries:
        127.0.0.0/8
       Live Restore Enabled: false
       Product License: Community Engine

-----> sigil version: 0.4.0
-----> herokuish version:
       herokuish: 0.5.0
       buildpacks:
         heroku-buildpack-multi     v1.0.0
         heroku-buildpack-ruby      v200
         heroku-buildpack-nodejs    v142
         heroku-buildpack-clojure   v83
         heroku-buildpack-python    v149
         heroku-buildpack-java      v63
         heroku-buildpack-gradle    v28
         heroku-buildpack-scala     v84
         heroku-buildpack-play      v26
         heroku-buildpack-php       v151
         heroku-buildpack-go        v104
         buildpack-nginx            v10
-----> dokku version: dokku version 0.17.9
-----> dokku plugins:
       plugn: 0.3.2
         00_dokku-standard    0.17.9 enabled    dokku core standard plugin
         20_events            0.17.9 enabled    dokku core events logging plugin
         app-json             0.17.9 enabled    dokku core app-json plugin
         apps                 0.17.9 enabled    dokku core apps plugin
         build-env            0.17.9 enabled    dokku core build-env plugin
         buildpacks           0.17.9 enabled    dokku core buildpacks plugin
         certs                0.17.9 enabled    dokku core certificate management plugin
         checks               0.17.9 enabled    dokku core checks plugin
         common               0.17.9 enabled    dokku core common plugin
         config               0.17.9 enabled    dokku core config plugin
         docker-options       0.17.9 enabled    dokku core docker-options plugin
         domains              0.17.9 enabled    dokku core domains plugin
         enter                0.17.9 enabled    dokku core enter plugin
         git                  0.17.9 enabled    dokku core git plugin
         logs                 0.17.9 enabled    dokku core logs plugin
         network              0.17.9 enabled    dokku core network plugin
         nginx-vhosts         0.17.9 enabled    dokku core nginx-vhosts plugin
         plugin               0.17.9 enabled    dokku core plugin plugin
         postgres             1.8.1 enabled    dokku postgres service plugin
         proxy                0.17.9 enabled    dokku core proxy plugin
         ps                   0.17.9 enabled    dokku core ps plugin
         repo                 0.17.9 enabled    dokku core repo plugin
         resource             0.17.9 enabled    dokku core resource plugin
         scheduler-docker-local 0.17.9 enabled    dokku core scheduler-docker-local plugin
         shell                0.17.9 enabled    dokku core shell plugin
         ssh-keys             0.17.9 enabled    dokku core ssh-keys plugin
         storage              0.17.9 enabled    dokku core storage plugin
         tags                 0.17.9 enabled    dokku core tags plugin
         tar                  0.17.9 enabled    dokku core tar plugin
         trace                0.17.9 enabled    dokku core trace plugin

**Comment by mfish0005** [comment]
@josegonzalez I tried the 3 methods listed in my first comment.  Which of those would you like to see an example of?  I.E. which of those methods sounds like the best solution for this use case?

Or should I just put together an example of the problem itself without my attempted solutions?

**Comment by josegonzalez** [comment]
Seems like you're running nginx inside of your container. The nginx config for _that_ nginx needs to be modified, not the one in Dokku. The host's nginx doesn't have access to the disk inside of your container, so `try_files` basically does nothing.

**Comment by mfish0005** [comment]
Oh I see.  My dockerfile is just using the base nginx image.  Should I create a custom nginx image?  Or can I just add an `nginx.conf` file to my project root and then add something like `COPY nginx.conf etc/nginx/nginx.conf` to my Dockerfile?

I realize this has nothing to do with Dokku.  Thanks for your help!

**Comment by josegonzalez** [comment]
Yeah that's basically it. Feel free to join us on our slack/irc (they are linked) with further questions.

**Comment by mfish0005** [comment]
I figured it out.  For anyone else that has this problem this is why: you must edit the YOUR container's nginx.  Not Dokku's nginx.

An example of how to fix it:
1. Create a local nginx.conf file in your project's root directory
2. Copy the default nginx config into it (E.G. https://github.com/oren/alpine-nginx/blob/master/nginx/default.conf)
3. Edit the default config to have whatever you want in it.  Mine looks like this:
```
server {
  location / {
      root /usr/share/nginx/html;
     # this is the line I added which makes nginx play nicely with Angular's router
      try_files $uri $uri/ /index.html;
  }

  location /item {
      alias /usr/share/nginx/html;
  }

  location /post {
      alias /usr/share/nginx/html;
  }
}
```
4. Add this to your Dockerfile
 ```
FROM nginx:alpine

RUN rm /etc/nginx/conf.d/default.conf

COPY nginx.conf etc/nginx/conf.d/default.conf
```

**Comment by josegonzalez** [comment]
Note: not all apps will have their own nginx. In this case, the user's docker setup has one, hence the need to modify things there.

**Comment by mfish0005** [comment]
Another note:  if you are only running a single dokku app `example.com/home` the above solution works great.  Unfortunately I also have an `api.example.com`.  Somehow I can access all of `example.com`'s routes from `api.example.com`.  

E.G. `example.com` has an about page accessible from `example.com/about`.  Somehow you can access that about page from `api.example.com/about` and this is causing errors with the API.

I'm not sure what the problem is but it seems to be preventing me from configuring SSL properly.  Will report back when I know more.

**Comment by pickettd** [comment]
I thought I should post a note here too because this issue thread was one of the top hits that I found while Googling for the problem I was having. I'm using create-react-app and the react-router with Dokku and was getting a 404 when I tried reloading the browser on http://my-app/dashboard. Like the original poster - I thought that I needed to modify application-level nginx in Dokku for try_files.

In my case however, my static files are being served by node/express in my app - so instead of doing anything with nginx, I needed to add something like this:
`app.get('*', (req,res) =>{
  res.sendFile(path.join(__dirname+'/client/build/index.html'));
});` in my server.js

Just thought I should post in case it helps others realize what is happening
